cpp/src/groupby/hash/aggregation_requests.cpp

```diff
@@ -96,64 +99,92 @@ gdf_column* compute_average(gdf_column sum, gdf_column count,
 
 table compute_original_requests(
     std::vector<AggRequestType> const& original_requests,
-    std::vector<AggRequestType> const& simple_requests, table simple_outputs,
-    cudaStream_t stream) {
-  // Maps the requested simple aggregation to the resulting output column
-  std::map<AggRequestType, gdf_column*> simple_requests_to_outputs;
+    std::vector<SimpleAggRequestCounter> const& simple_requests,
+    table simple_outputs, cudaStream_t stream) {
+  // Maps the requested simple aggregation to a resulting output column paired
+  // with a counter of how many times said column is needed
+  std::map<AggRequestType, std::pair<gdf_column*, gdf_size_type>>
+      simple_requests_to_outputs;
 
   for (std::size_t i = 0; i < simple_requests.size(); ++i) {
     CUDF_EXPECTS(simple_outputs.get_column(i) != nullptr,
                  "Missing output column[" + std::to_string(i) + "]");
 
-    simple_requests_to_outputs[simple_requests[i]] =
-        simple_outputs.get_column(i);
+    simple_requests_to_outputs[simple_requests[i].first] = {
+        simple_outputs.get_column(i), simple_requests[i].second};
   }
 
-  std::vector<gdf_column*> final_value_columns;
+  // Calculate processing order so that the compound operators are
+  // processed first
+  std::vector<gdf_size_type> processing_order(original_requests.size());
+  std::iota(processing_order.begin(), processing_order.end(), 0);
```

Comment:
I see what you mean about still needing to sort the requests to avoid unnecessary copies. 

How about instead of relying on the enum ordering for the sort, we do a 2 pass approach? 

Do a first pass through the `original_requests` and compute the compound requests (decrementing the counts of the associated `simple_requests` as necessary without making any copies).

Then do a second pass, of which only simple requests should remain. For each simple request, make copies and decrement the count until the `count == 1`, at which point, no copy is made, just use the computed result. Then remove the request.

Reply:
done